### PR TITLE
Rpi 4.19.y unicam driver

### PR DESCRIPTION
**Pull request by 6by9**
Cherry-picking all the Unicam driver and "supported" source subdevice driver fixes from 4.14 to 4.19, with minor fixups due to API changes.
TC358743 and ADV7282 both tested and working.

### CODE REVIEW

File: arch/arm/boot/dts/bcm2708-rpi.dtsi [hunk 1]
```diff
@@ -161,3 +161,11 @@ sdhost_pins: &sdhost_gpio48 {
...
 	status = "disabled";
 };
+
+&csi0 {
+	power-domains = <&power RPI_POWER_DOMAIN_UNICAM0>;
+};
+
+&csi1 {
+	power-domains = <&power RPI_POWER_DOMAIN_UNICAM1>;
+};
```
Comment (lategoodbye) [inline arch/arm/boot/dts/bcm2708-rpi.dtsi:171]: Not sure this is intended, but i expected this in the downstream DTS commit

File: arch/arm/boot/dts/bcm283x.dtsi [hunk 2]
```diff
@@ -565,6 +565,34 @@
...
+			compatible = "brcm,bcm2835-unicam";
+			reg = <0x7e800000 0x800>,
+			      <0x7e802000 0x4>;
+			interrupts = <2 6>;
+			clocks = <&clocks BCM2835_CLOCK_CAM0>;
+			clock-names = "lp";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			#clock-cells = <1>;
+
```
Comment (lategoodbye) [inline arch/arm/boot/dts/bcm283x.dtsi]: For upstreaming these empty lines should be avoided

File: arch/arm64/boot/dts/broadcom/bcm283x-rpi-csi1-2lane.dtsi [hunk 3]
```diff
@@ -0,0 +1 @@
+../../../../arm/boot/dts/bcm283x-rpi-csi1-2lane.dtsi
```
Comment (lategoodbye) [inline arch/arm64/boot/dts/broadcom/bcm283x-rpi-csi1-2lane.dtsi:1]: Just a hint, the CM3 will be added in 4.20 so the other csi variants could be added too

File: arch/arm/boot/dts/bcm283x.dtsi [hunk 4]
```diff
@@ -565,6 +565,34 @@
 			status = "disabled";
 		};
 
+		csi0: csi0@7e800000 {
```
Comment (lategoodbye) [inline arch/arm/boot/dts/bcm283x.dtsi]: The node name should describe the function like csi, so my suggestion would be:
csi0: csi@7e800000

File: arch/arm/boot/dts/bcm2708-rpi.dtsi (hunk 1 above)
Comment (6by9) [reply arch/arm/boot/dts/bcm2708-rpi.dtsi:171]: Yup. Done.

File: arch/arm/boot/dts/bcm283x.dtsi (hunk 4 above)
Comment (6by9) [reply arch/arm/boot/dts/bcm283x.dtsi]: Done

File: arch/arm/boot/dts/bcm283x.dtsi (hunk 2 above)
Comment (6by9) [reply arch/arm/boot/dts/bcm283x.dtsi]: Removed.

File: arch/arm64/boot/dts/broadcom/bcm283x-rpi-csi1-2lane.dtsi (hunk 3 above)
Comment (6by9) [reply arch/arm64/boot/dts/broadcom/bcm283x-rpi-csi1-2lane.dtsi:1]: They could, but creating files that are unused and untested tends to cause more upset than having to link to them later on.

**Review by 6by9 — COMMENTED** [review]
Updated patch set incoming.

**Comment by 6by9** [comment]
PR updated based on @lategoodbye 's comments.

File: arch/arm/boot/dts/bcm283x.dtsi (hunk 2 above)
Comment (lategoodbye) [reply arch/arm/boot/dts/bcm283x.dtsi]: Maybe there was a misunderstanding. I wasn't talking about the cells properties. I meant really the empty lines in the csi0 and csi1 node.

File: arch/arm64/boot/dts/broadcom/bcm283x-rpi-csi1-2lane.dtsi (hunk 3 above)
Comment (lategoodbye) [reply arch/arm64/boot/dts/broadcom/bcm283x-rpi-csi1-2lane.dtsi:1]: Okay

File: arch/arm/boot/dts/bcm283x.dtsi (hunk 2 above)
Comment (6by9) [reply arch/arm/boot/dts/bcm283x.dtsi]: OK. Although I see a fair number of blank lines in gpio, i2s, dsi0, dsi1.
Another update incoming.

**Comment by 6by9** [comment]
Rebased, @lategoodbye's last comments actioned, and retested with both TC358743 and ADV7282M.
@pelwell anything else you want?

**Pull request merged**